python/cuda_parallel/cuda/parallel/experimental/_iterators.py

```diff
@@ -0,0 +1,428 @@
+        thisty = numba.types.CPointer(ntype)
+        self.val = _ctypes_type_given_numba_type(ntype)(val)
+        self.ntype = ntype
+        self.prefix = "constant_" + ntype.name
+        self.ltoirs = [
+            _ncc(
+                f"{self.prefix}_advance",
+                constant_advance,
+                numba.types.void(thisty, _DISTANCE_NUMBA_TYPE),
+            ),
+            _ncc(f"{self.prefix}_dereference", constant_dereference, ntype(thisty)),
+        ]
+
+    @property
+    def state_c_void_p(self):
+        return ctypes.cast(ctypes.pointer(self.val), ctypes.c_void_p)
+
+    @property
+    def size(self):
+        return self.ntype.bitwidth // 8
+
+    @property
+    def alignment(self):
+        return self.size
+
+
+def count_advance(this, diff):
+    this[0] += diff
+
+
+def count_dereference(this):
+    return this[0]
+
+
+class CountingIterator:
+    def __init__(self, count, ntype):
+        thisty = numba.types.CPointer(ntype)
+        self.count = _ctypes_type_given_numba_type(ntype)(count)
+        self.ntype = ntype
+        self.prefix = "count_" + ntype.name
+        self.ltoirs = [
+            _ncc(
+                f"{self.prefix}_advance",
+                count_advance,
+                numba.types.void(thisty, _DISTANCE_NUMBA_TYPE),
+            ),
+            _ncc(f"{self.prefix}_dereference", count_dereference, ntype(thisty)),
+        ]
+
+    @property
+    def state_c_void_p(self):
+        return ctypes.cast(ctypes.pointer(self.count), ctypes.c_void_p)
+
+    @property
+    def size(self):
+        return self.ntype.bitwidth // 8
+
+    @property
+    def alignment(self):
+        return self.size
+
+
+class TransformIteratorImpl:
+    def __init__(
+        self,
+        it,
+        op,
+        op_return_ntype,
+        transform_advance,
+        transform_dereference,
+        op_abi_name,
+    ):
+        self.it = it
+        self.ntype = op_return_ntype
+        self.prefix = f"transform_{it.prefix}_{op.__name__}"
+        self.ltoirs = it.ltoirs + [
+            _ncc(
+                f"{self.prefix}_advance",
+                transform_advance,
+                numba.types.void(
+                    numba.types.CPointer(numba.types.char), _DISTANCE_NUMBA_TYPE
+                ),
+            ),
+            _ncc(
+                f"{self.prefix}_dereference",
+                transform_dereference,
+                op_return_ntype(numba.types.CPointer(numba.types.char)),
+            ),
+            # ATTENTION: NOT op_caller here! (see issue #3064)
+            _ncc(op_abi_name, op, op_return_ntype(it.ntype)),
+        ]
+
+    @property
+    def state_c_void_p(self):
+        return self.it.state_c_void_p
+
+    @property
+    def size(self):
+        return self.it.size  # TODO fix for stateful op
+
+    @property
+    def alignment(self):
+        return self.it.alignment  # TODO fix for stateful op
+
+
+def TransformIterator(op, it, op_return_ntype):
```

Comment:
I don't think we should require the `op_return_ntype` here. Numba should in theory have everything it needs to infer the return type when compiling `op`. 

`cuda.compile` returns both the LTOIR as well as the inferred return type, which we seem to be discarding in `extract_ctypes_ltoirs`.

Are we able to use the numba inferred return type and not require it from the user?

If not, it might be because numba doesn't have enough typing information. If that is the case, it will be fixed as part of https://github.com/NVIDIA/cccl/issues/3064 by defining numba types corresponding to all of our `Iterator`types.